File: shortest_distance_I.py
# approach - update indices when we find words 
# Return min distance recorded so far

# Time - O(N)
# Space - O(1)

class Solution:
    def shortestDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        
        idx1 = -1
        idx2 = -1
        result = float("inf")
        
        for idx, word in enumerate(wordsDict):
            
            if word == word1:
                idx1 = idx
                
            elif word == word2:
                idx2 = idx
            
            
            if idx1 != -1 and idx2 != -1:
                result = min(result, abs(idx1-idx2))
                
        return result
                
                
        

File: ShortestDistanceI.java
// Time: O(N) | Space: O(1)

class Solution {
    public int shortestDistance(String[] wordsDict, String word1, String word2) {
        int p1 = -1;
        int p2 = -1;
        int min = Integer.MAX_VALUE;
        for(int i=0;i<wordsDict.length;i++) {
            if(word1.equals(wordsDict[i])) p1 = i;
            if(word2.equals(wordsDict[i])) p2 = i;
            if(p1 != -1 && p2 != -1)
                min = Math.min(min, Math.abs(p1-p2));
        }
        return min;
    }
}